src/crimson/os/seastore/linked_tree_node.h

```diff
@@ -0,0 +1,995 @@
+    take_children_from_prior_instance();
+    assert(me.validate_stable_children());
+  }
+
+  // children are considered stable if any of the following case is true:
+  // 1. The child extent is absent in cache
+  // 2. The child extent is (data) stable
+  //
+  // For reserved mappings, the return values are undefined.
+  bool _is_child_stable(
+    Transaction &t,
+    uint16_t pos,
+    node_key_t key,
+    bool data_only = false) const {
+    ceph_assert(cache);
+    auto &me = down_cast();
+    assert(key == me.iter_idx(pos).get_key());
+    auto child = this->children[pos];
+    if (is_reserved_ptr(child)) {
+      return true;
+    } else if (is_valid_child_ptr(child)) {
+      assert(dynamic_cast<CachedExtent*>(child)->is_logical());
+      assert(
+	dynamic_cast<CachedExtent*>(child)->is_pending_in_trans(t.get_trans_id())
+	|| me.is_stable_written());
+      if (data_only) {
+	return cache->is_viewable_extent_data_stable(
+	  t, dynamic_cast<CachedExtent*>(child));
+      } else {
+	return cache->is_viewable_extent_stable(
+	  t, dynamic_cast<CachedExtent*>(child));
+      }
+    } else if (me.is_pending()) {
+      auto key = me.iter_idx(pos).get_key();
+      auto &sparent = me.get_stable_for_key(key);
+      auto spos = sparent.lower_bound(key).get_offset();
+      auto child = sparent.children[spos];
+      if (is_valid_child_ptr(child)) {
+	assert(dynamic_cast<CachedExtent*>(child)->is_logical());
+	if (data_only) {
+	  return cache->is_viewable_extent_data_stable(
+	    t, dynamic_cast<CachedExtent*>(child));
+	} else {
+	  return cache->is_viewable_extent_stable(
+	    t, dynamic_cast<CachedExtent*>(child));
+	}
+      } else {
+	return true;
+      }
+    } else {
+      return true;
+    }
+  }
+
+  parent_tracker_t<T>* my_tracker = nullptr;
+private:
+  T& down_cast() {
+    return *static_cast<T*>(this);
+  }
+  const T& down_cast() const {
+    return *static_cast<const T*>(this);
+  }
+
+  std::vector<BaseChildNode<T, node_key_t>*> children;
+  std::set<TCachedExtentRef<T>, Comparator> copy_sources;
+  uint16_t capacity = 0;
+  uint16_t num_children = 0;
+
+  // copy dests points from a stable node back to its pending nodes
+  // having copy sources at the same tree level, it serves as a two-level index:
+  // transaction-id then node-key to the pending node.
+  //
+  // The copy dest pointers must be symmetric to the copy source pointers.
+  //
+  // copy_dests_t will be automatically unregisterred upon transaction destruction,
+  // see Transaction::views
+  struct copy_dests_t : trans_spec_view_t {
+    std::set<TCachedExtentRef<T>, Comparator> dests_by_key;
+    copy_dests_t(Transaction &t) : trans_spec_view_t{t.get_trans_id()} {}
+    ~copy_dests_t() {
+      LOG_PREFIX(~copy_dests_t);
+      SUBTRACE(seastore_fixedkv_tree, "copy_dests_t destroyed");
+    }
+  };
+
+  using trans_view_set_t = trans_spec_view_t::trans_view_set_t;
+  trans_view_set_t copy_dests_by_trans;
+  CacheProxyRef cache;
```

Comment:
Following the above comment, is it really necessary to hold a `CacheProxyRef`?  Fundamentally it is still a `Cache&`.

Can `CacheProxy&` be passed to the related methods instead?

Reply:
Ah, yes.